OrbitQt/SourceCodeView.h

```diff
@@ -0,0 +1,86 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_QT_SOURCE_CODE_VIEW_H_
+#define ORBIT_QT_SOURCE_CODE_VIEW_H_
+
+#include <QObject>
+#include <QVector>
+
+#include "WebEngine/Dialog.h"
+
+namespace orbit_qt {
+
+/* SourceCodeView is a wrapper around web_engine::Dialog for view highlighted code.
+
+   The class exposes three properties (title, source_code, and language) which can
+   be read by the corresponding web part of SourceCodeView. Use SetSourceCode
+   to change any of these three values. The web part will also be notified and
+   will reload accordingly. "language" is an identifier which needs to be recognized
+   by PrismJS to identify the syntax highlighting language. Currently we support
+   "c", "cpp", and "x86asm". Any other value will lead to no highlighting.
+
+   Example:
+   QWebEngineProfile profile;
+   SourceCodeView view{&profile};
+   view.SetSourceCode("example.cpp", "int main() {}", "cpp");
+   view.exec(); // Will open the dialog and block until closed.
+*/
+class SourceCodeView : public QObject {
+  Q_OBJECT
+
+ public:
+  SourceCodeView(QWebEngineProfile* web_engine_profile, QObject* parent = nullptr);
+
+  Q_PROPERTY(QString title MEMBER title_ NOTIFY sourceCodeChanged)
+  Q_PROPERTY(QString source_code MEMBER source_code_ NOTIFY sourceCodeChanged)
+  Q_PROPERTY(QString language MEMBER language_ NOTIFY sourceCodeChanged)
+  Q_PROPERTY(bool line_numbers_enabled MEMBER line_numbers_enabled_ NOTIFY sourceCodeChanged)
+  Q_PROPERTY(bool heatmap_enabled MEMBER heatmap_enabled_ NOTIFY sourceCodeChanged)
+  Q_PROPERTY(QVector<int> hit_counts MEMBER hit_counts_ NOTIFY sourceCodeChanged)
+  Q_PROPERTY(QVector<double> hit_ratios MEMBER hit_ratios_ NOTIFY sourceCodeChanged)
+  Q_PROPERTY(QVector<double> total_hit_ratios MEMBER total_hit_ratios_ NOTIFY sourceCodeChanged)
+
+  void SetSourceCode(QString title, QString new_source, QString language);
+  void SetHeatmapData(QVector<int> hit_counts, QVector<double> hit_ratios,
+                      QVector<double> total_hit_ratios);
+
+  void SetLineNumbersEnabled(bool enabled) {
+    if (enabled != line_numbers_enabled_) {
+      line_numbers_enabled_ = enabled;
+      emit sourceCodeChanged();
+    }
+  }
+  void SetHeatmapEnabled(bool enabled) {
+    if (enabled != heatmap_enabled_) {
+      heatmap_enabled_ = enabled;
+      emit sourceCodeChanged();
+    }
+  }
+  int exec();
+
+  Q_INVOKABLE void CloseDialog() { dialog_.close(); }
+
+  Q_INVOKABLE void SetWindowTitle(QString title) { dialog_.setWindowTitle(std::move(title)); }
+
+  int GetWebSocketServerListenPort() const { return dialog_.GetWebSocketServer()->serverPort(); }
+
+ signals:
+  void sourceCodeChanged();
+
+ private:
+  web_engine::Dialog dialog_;
+
+  QString title_;
+  QString source_code_;
+  QString language_;
+  bool line_numbers_enabled_ = false;
+  bool heatmap_enabled_ = false;
+  QVector<int> hit_counts_;
```

Comment:
Shouldn't we be able to store the DisassemblyReport here, and then just have functions for the javascript part, to retrieve these numbers?

Reply:
I changed that for two reasons:
1. I replaced function calls by Q_PROPERTY to not have 3 function calls per line which lead to a lot of webchannel messages. Pushing all values in one message to the JavaScript context is a lot faster.
2. When I made that line number mangling (which you already pointed out) I wanted to have all the code which is messing with the line numbers in one place. That's why the DisassemblyReport is not part of this class anymore.

I think we should talk about this on how to move forward here. I'm happy to change things for better. :smile:

Reply:
I will resolve this for now. I think you need to have a fresh look at this anyway.